File: python/src/mechbase/resources/measurements.py
from __future__ import annotations

import io
import json
from pathlib import Path
from typing import Any, BinaryIO

from ..models import BatchResult, FileAttachment, Measurement
from ._base import InstallationScoped, _paginated


class Measurements(InstallationScoped):
    def create(
        self,
        *,
        point_id: int,
        data: dict,
        notes: str = "",
        session_id: str | None = None,
        timestamp: str | None = None,
        file: str | Path | BinaryIO | None = None,
        idempotency_key: str | None = None,
    ) -> Measurement:
        payload: dict[str, Any] = {
            "measurement_point_id": point_id,
            "data": data,
            "notes": notes,
        }
        if session_id:
            payload["session_id"] = session_id
        if timestamp:
            payload["timestamp"] = timestamp

        headers = {"X-Idempotency-Key": idempotency_key} if idempotency_key else None

        if file is None:
            body = self._http.request(
                "POST", self._path("/measurements/"), json=payload, headers=headers
            )
            return Measurement.from_dict(body)

        fh = file
        opened = False
        if isinstance(file, (str, Path)):
            fh = open(file, "rb")  # noqa: SIM115
            opened = True
        try:
            body = self._http.request(
                "POST",
                self._path("/measurements/upload/"),
                data={"payload": json.dumps(payload)},
                files={"file": (Path(getattr(fh, "name", "attachment")).name, fh)},
                headers=headers,
            )
        finally:
            if opened:
                fh.close()
        return Measurement.from_dict(body)

    def list_for_point(self, point_id: int, *, limit: int = 50, offset: int = 0,
                       created_from: str | None = None, created_to: str | None = None) -> list[Measurement]:
        return _paginated(
            self._http,
            self._path(f"/measurement-points/{point_id}/measurements/"),
            params={"limit": limit, "offset": offset,
                    "created_from": created_from, "created_to": created_to},
            parse=Measurement.from_dict,
        )

    def create_batch(self, items: list[dict]) -> BatchResult:
        body = self._http.request(
            "POST", self._path("/measurements/batch/"), json={"items": list(items)}
        )
        return BatchResult.from_dict(body)

    def add_file(
        self,
        measurement_uuid: str,
        *,
        file: str | Path | BinaryIO | bytes,
        filename: str = "attachment",
    ) -> FileAttachment:
        fh = file
        opened = False
        if isinstance(file, (str, Path)):
            filename = Path(file).name
            fh = open(file, "rb")  # noqa: SIM115
            opened = True
        elif isinstance(file, (bytes, bytearray)):
            fh = io.BytesIO(file)
        try:
            body = self._http.request(
                "POST",
                self._path(f"/measurements/{measurement_uuid}/files/"),
                files={"file": (filename, fh)},
            )
        finally:
            if opened:
                fh.close()
        return FileAttachment.from_dict(body)

    def iter_for_point(
        self,
        point_id: int,
        *,
        created_from: str | None = None,
        created_to: str | None = None,
        page_size: int = 100,
    ):
        """Yield every measurement for a point, following cursor pages."""
        cursor: str | None = ""
        path = self._path(f"/measurement-points/{point_id}/measurements/")
        while True:
            params = {
                "limit": page_size,
                "cursor": cursor,
                "created_from": created_from,
                "created_to": created_to,
            }
            params = {k: v for k, v in params.items() if v is not None}
            body = self._http.request("GET", path, params=params)
            for item in body["items"]:
                yield Measurement.from_dict(item)
            cursor = body.get("next_cursor")
            if not cursor:
                break

File: go/pkg/mechbase/measurements.go
package mechbase

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
)

// Measurements is the measurements resource for one installation.
type Measurements struct {
	client *Client
	iid    int
}

// CreateMeasurementInput is the payload for Measurements.Create.
type CreateMeasurementInput struct {
	PointID   int
	Data      map[string]any
	Notes     string
	SessionID string
	Timestamp string
	// File, if non-nil, switches the request to multipart upload.
	File     io.Reader
	FileName string
	// IdempotencyKey, when set, is sent as the X-Idempotency-Key header so the
	// server dedupes a retried create.
	IdempotencyKey string
}

// Create posts a single measurement. When File is nil it sends JSON; when set
// it sends multipart with the canonical "payload" JSON field.
func (m *Measurements) Create(ctx context.Context, in CreateMeasurementInput) (*Measurement, error) {
	payload := map[string]any{
		"measurement_point_id": in.PointID,
		"data":                 in.Data,
		"notes":                in.Notes,
	}
	if in.SessionID != "" {
		payload["session_id"] = in.SessionID
	}
	if in.Timestamp != "" {
		payload["timestamp"] = in.Timestamp
	}

	var out Measurement
	headers := idempotencyHeaders(in.IdempotencyKey)
	if in.File == nil {
		if err := m.client.doJSONH(ctx, http.MethodPost, pathf(m.iid, "/measurements/"), nil, headers, payload, &out); err != nil {
			return nil, err
		}
		return &out, nil
	}

	encoded, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("mechbase: marshal payload: %w", err)
	}
	name := in.FileName
	if name == "" {
		name = "attachment"
	}
	files := []filePart{{Field: "file", Filename: name, Reader: in.File}}
	fields := map[string]string{"payload": string(encoded)}
	if err := m.client.doMultipartH(ctx, pathf(m.iid, "/measurements/upload/"), headers, fields, files, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ListForPointOptions filters ListForPoint. Zero values are omitted.
type ListForPointOptions struct {
	Limit       int
	Offset      int
	CreatedFrom string
	CreatedTo   string
}

// ListForPoint returns one offset page of a point's history.
func (m *Measurements) ListForPoint(ctx context.Context, pointID int, opts ListForPointOptions) ([]Measurement, error) {
	q := url.Values{}
	addPaging(q, opts.Limit, opts.Offset)
	if opts.CreatedFrom != "" {
		q.Set("created_from", opts.CreatedFrom)
	}
	if opts.CreatedTo != "" {
		q.Set("created_to", opts.CreatedTo)
	}
	var env listEnvelope[Measurement]
	path := pathf(m.iid, fmt.Sprintf("/measurement-points/%d/measurements/", pointID))
	if err := m.client.doJSON(ctx, http.MethodGet, path, q, nil, &env); err != nil {
		return nil, err
	}
	return env.Items, nil
}

// MeasurementInput is one item in a measurement batch. Supply exactly one of
// PointID / PointExternalID. Nil fields are omitted.
type MeasurementInput struct {
	MeasurementPointID         *int           `json:"measurement_point_id,omitempty"`
	MeasurementPointExternalID *string        `json:"measurement_point_external_id,omitempty"`
	Data                       map[string]any `json:"data"`
	Notes                      *string        `json:"notes,omitempty"`
	SessionID                  *string        `json:"session_id,omitempty"`
	Timestamp                  *string        `json:"timestamp,omitempty"`
	ExternalID                 *string        `json:"external_id,omitempty"`
}

// CreateBatch pushes many measurements in one request.
func (m *Measurements) CreateBatch(ctx context.Context, items []MeasurementInput) (*BatchResult, error) {
	body := map[string]any{"items": items}
	var out BatchResult
	if err := m.client.doJSON(ctx, http.MethodPost, pathf(m.iid, "/measurements/batch/"), nil, body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// AddFile attaches a file to an existing measurement (multipart).
func (m *Measurements) AddFile(ctx context.Context, measurementUUID, filename string, file io.Reader) (*FileAttachment, error) {
	if filename == "" {
		filename = "attachment"
	}
	files := []filePart{{Field: "file", Filename: filename, Reader: file}}
	var out FileAttachment
	if err := m.client.doMultipart(ctx, pathf(m.iid, fmt.Sprintf("/measurements/%s/files/", measurementUUID)), nil, files, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// IterOptions configures IterForPoint.
type IterOptions struct {
	CreatedFrom string
	CreatedTo   string
	PageSize    int
}

// IterForPoint streams the full history for a point, following cursor pages.
// The callback is invoked per measurement; return an error to stop.
func (m *Measurements) IterForPoint(ctx context.Context, pointID int, opts IterOptions, fn func(Measurement) error) error {
	cursor := ""
	path := pathf(m.iid, fmt.Sprintf("/measurement-points/%d/measurements/", pointID))
	for {
		q := url.Values{}
		q.Set("cursor", cursor)
		if opts.PageSize > 0 {
			q.Set("limit", fmt.Sprintf("%d", opts.PageSize))
		}
		if opts.CreatedFrom != "" {
			q.Set("created_from", opts.CreatedFrom)
		}
		if opts.CreatedTo != "" {
			q.Set("created_to", opts.CreatedTo)
		}
		var env cursorEnvelope[Measurement]
		if err := m.client.doJSON(ctx, http.MethodGet, path, q, nil, &env); err != nil {
			return err
		}
		for _, item := range env.Items {
			if err := fn(item); err != nil {
				return err
			}
		}
		if env.NextCursor == nil || *env.NextCursor == "" {
			return nil
		}
		cursor = *env.NextCursor
	}
}
